fix: Won't build tests & docs if not top project.

cnihelton opened the pull request:
## References

 - Issue: Fixes #4.

## What changes

Moves building docs and tests into the same section as packaging, to run only if this is the top level project.

## Checklist

- [x] Bug fix (change which fixes issues)
- [ ] New feature (change which introduces functionality)
- [ ] Breaking change (that causes any existing test to change, e.g. API / ABI change)
- [x] I have read the [contributing guidelines](CONTRIBUTING.md).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

CarlosNihelton committed:
fix: Won't build tests & docs if not top project.

Makes it easier for newcomers to use the library
Also makes it easier for package maintainers.

m8mble commented:
Thanks for the PR. With your change, users can no longer enable tests when they use `clean-test` in a sub-directory - even if they want to. Professional CMake (Ed 15) suggests the following:

```cmake
option(CLEANTEST_TEST "Build tests for the framework." ${PROJECT_IS_TOP_LEVEL})

# [ ... ]

if (CLEANTEST_TEST)
  if (PROJECT_IS_TOP_LEVEL)
    enable_testing()
  endif()
  add_subdirectory(test)
endif()
if (PROJECT_IS_TOP_LEVEL)
  # [ ... ] (packaging
else()
  mark_as_advanced(CLEANTEST_TEST)
endif()
```
Would that meet your expectations?

Similar considerations apply to the `docs` folder. If docs are annoying by default, I'd prefer to rather exclude related targets from `all` but keeping the `add_subdirectory(docs)` unconditionally.

Commit:
Uses PROJECT_IS_TOP_LEVEL variable

Following 'Professional CMake' recommendation
Requires CMake 3.21, though.

CarlosNihelton commented:
> Would that meet your expectations?

Absolutely! `PROJECT_IS_TOP_LEVEL` requires CMake 3.21 or higher, though. 

> 
> Similar considerations apply to the `docs` folder. If docs are annoying by default, I'd prefer to rather exclude related targets from `all` but keeping the `add_subdirectory(docs)` unconditionally.

Indeed, the `add_custom_target` doesn't affect the `ALL` target by default, so since you didn't add the  `ALL` flag when adding the doc targets, we are fine with unconditionally adding the docs sub-directory.

m8mble commented on CMakeLists.txt:
```diff
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: BSL-1.0
 
-cmake_minimum_required(VERSION 3.17)
+cmake_minimum_required(VERSION 3.21)
```
Thanks for spotting this. Let me suggest
```cmake
set(CLEANTEST_IS_TOP_LEVEL OFF)
if (CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
    set(CLEANTEST_IS_TOP_LEVEL ON)
endif()
```
(and replacing `PROJECT_IS_TOP_LEVEL => CLEANTEST_IS_TOP_LEVEL`) as a workaround in order to stay with the `cmake` version.

m8mble commented on CMakeLists.txt:
```diff
@@ -285,4 +287,6 @@ if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
     )
     include(CPack)
+else()
+  mark_as_advanced(CLEANTEST_TEST)
```
:+1:

m8mble reviewed (commented):
Thanks for checking the minimum required `cmake` version.

Commit:
Avoids bumping CMake minimum version

By replacing PROJECT_IS_TOP_LEVEL with
local var set by comparing the current with the top level CMake dir.

m8mble reviewed (approved):
LGTM

m8mble commented:
Thank you very much for fixing this.

CarlosNihelton commented:
Thank you very much for this awesome library! I'd like to see it widespread.
